BTrDB.cs

```diff
@@ -1,131 +1,732 @@
         {
-            throw new NotImplementedException("not implemented");
+            V5Api.NearestParams p = new V5Api.NearestParams
+            {
+                Uuid = ByteString.CopyFrom(id.ToByteArray()),
+                Time = time,
+                Backward = backward,
+            };
+            V5Api.NearestResponse res = db.client.Nearest(p);
+            if (res.Stat != null)
+            {
+                throw new BTrDBException(res.Stat);
+            }
+            return (new RawPoint(res.Value), res.VersionMajor, res.VersionMinor);
         }
 
-        public (IEnumerable<ChangedRange>, ulong, ulong) Changes(ulong fromVersion, ulong toVersion, ushort depth)
+        public ICursor<ChangedRange> Changes(ulong fromVersion, ulong toVersion, ushort depth)
         {
-            throw new NotImplementedException("not implemented");
+            V5Api.ChangesParams p = new V5Api.ChangesParams
+            {
+                Uuid = ByteString.CopyFrom(id.ToByteArray()),
+                FromMajor = fromVersion,
+                ToMajor = toVersion,
+                Resolution = depth,
+            };
+            CancellationTokenSource ctx = new CancellationTokenSource();
+            var res = db.client.Changes(p, cancellationToken:ctx.Token).ResponseStream;
+            return new Cursor<ChangedRange, V5Api.ChangesResponse>(ctx, res, (V5Api.ChangesResponse src) =>
+            {
+                if (src.Stat != null) throw new BTrDBException(src.Stat);
+                return from r in src.Ranges select new ChangedRange(r);
+            });
         }
 
         public (ulong, ulong) Insert(IEnumerable<RawPoint> data)
         {
-            throw new NotImplementedException("not implemented");
+            const int batchsz = 50000;
+            V5Api.InsertParams parms = new V5Api.InsertParams 
+            { 
+                Uuid = ByteString.CopyFrom(id.ToByteArray()) 
+            };
+            ulong lastmaj =0, lastmin =0;
+            foreach (RawPoint p in data)
+            {
+                parms.Values.Add(new V5Api.RawPoint { Time = p.Time, Value = p.Value });
+                if (parms.Values.Count >= batchsz)
+                {
+                    var res = db.client.Insert(parms);
+                    if (res.Stat != null) throw new BTrDBException(res.Stat);
+                    lastmaj = res.VersionMajor;
+                    lastmin = res.VersionMinor;
+                    parms = new V5Api.InsertParams
+                    { 
+                        Uuid = ByteString.CopyFrom(id.ToByteArray())
+                    };
+                }
+            }
+            if (parms.Values.Count > 0)
+            {
+                var res = db.client.Insert(parms);
+                if (res.Stat != null) throw new BTrDBException(res.Stat);
+                lastmaj = res.VersionMajor;
+                lastmin = res.VersionMinor;
+            }
+            return (lastmaj, lastmin);
```

Comment:
+1 going to be nice to send any size iterable at this and have it batch for us. We may want to put suggested batch size as a constant outside the func in case we want to write concurrent inserts from PI

Reply:
Ok, will move it